src/main/java/io/vertx/httpproxy/impl/CacheImpl.java

```diff
@@ -1,23 +1,51 @@
 package io.vertx.httpproxy.impl;
 
+import io.vertx.core.Future;
+import io.vertx.core.Vertx;
 import io.vertx.httpproxy.cache.CacheOptions;
 import io.vertx.httpproxy.spi.cache.Cache;
+import io.vertx.httpproxy.spi.cache.Resource;
 
-import java.util.LinkedHashMap;
-import java.util.Map;
+import java.util.*;
 
 /**
  * Simplistic implementation.
  */
-public class CacheImpl<K, V> extends LinkedHashMap<K, V> implements Cache<K, V> {
+public class CacheImpl implements Cache {
 
   private final int maxSize;
+  private final Map<String, Resource> data;
 
   public CacheImpl(CacheOptions options) {
     this.maxSize = options.getMaxSize();
+    this.data = Collections.synchronizedMap(new LinkedHashMap<>() {
+      @Override
+      protected boolean removeEldestEntry(Map.Entry<String, Resource> eldest) {
+        return size() > maxSize;
+      }
+    });
   }
 
-  protected boolean removeEldestEntry(Map.Entry eldest) {
-    return size() > maxSize;
+
+  @Override
+  public Future<Void> put(String key, Resource value) {
+    data.put(key, value);
+    return Future.succeededFuture();
+  }
+
+  @Override
+  public Future<Resource> get(String key) {
+    return Future.succeededFuture(data.get(key));
+  }
+
+  @Override
+  public Future<Void> remove(String key) {
+    data.remove(key);
+    return Future.succeededFuture();
+  }
+
+  @Override
+  public Future<Void> close() {
+    return Future.succeededFuture();
```

Comment:
Personally I also think there's no need, but when I called the `vertx.createSharedResource()` in `ReverseProxy#newCache`, there's a `closeFuture` that I don't know how to deal with  it.

Maybe you could check if the following would work? I think I should change to this version:

```java
  // ReverseProxy#newCache
  public Cache newCache(CacheOptions options, Vertx vertx) {
    if (options.getShared()) {
      CloseFuture closeFuture = new CloseFuture();
      return ((VertxInternal) vertx).createSharedResource("__vertx.shared.proxyCache", options.getName(), closeFuture, (cf_) -> {
        Cache cache = new CacheImpl(options);
        return cache;
      });
    }
    return new CacheImpl(options);
  }
```

Reply:
The close future is not bound to anything, neither Vert.x nor the context. Take a look at how Vert.x creates shared http clients, or shared Vert.x SQL Clients, shared Kafka Producers

Reply:
Is this correct then?

```java
  public Cache newCache(CacheOptions options, Vertx vertx) {
    if (options.getShared()) {
      VertxInternal vx = (VertxInternal) vertx;
      CloseFuture closeFuture = new CloseFuture();
      Cache cache = vx.createSharedResource("__vertx.shared.proxyCache", options.getName(), closeFuture, (cf_) -> {
        return new CacheImpl(options);
      });
      ContextInternal ctx = vx.getContext();
      if (ctx != null) {
        ctx.addCloseHook(closeFuture);
      } else {
        vx.addCloseHook(closeFuture);
      }
      return cache;
    }
    return new CacheImpl(options);
  }
```

So now I know the callback (the `cf_` one) doesn't need to close anything because there's nothing to close for the Cache interface; and for the outside `closeFuture`, my current understanding is that by calling `closeFuture.close()` means declaring "giving up the reference holding" of this shared resource and reduce the reference counter (`Hook.count`) by 1. 

Correct me if I'm wrong: the outside `closeFuture` still should be managed because we want to maintain the reference counter correct, and when the `vertx.close()` is called, the counter will be cleared to 0 even if `newCache` being called multiple times (because `addCloseHook` is registered multiple times as well).